lib/puppet/parser/compiler.rb

```diff
@@ -556,7 +556,15 @@ def evaluate_ast_node
     # of resources.
     resource = astnode.ensure_in_catalog(topscope)
 
-    resource.evaluate
+    # Evaluate the AST node unless it is the 'default' node. The 'default' node will
+    # then be evaluated together with other ENC added classes when #eval_definitions
+    # is called, giving those classes a chance to override definitions that the
+    # default node is using.
+    #
+    # If the 'default' node is evaluated right here, then any definition that it
+    # instantiates, will be evaluated during the #eval_definitions phase and
+    # before evaluation of classes included by ENC.
+    resource.evaluate unless astnode.name == 'default'
 
```

Comment:
This is scary:
* I think this means that the problem still exists if the node is matched by a specific node expression. (i.e. if the test in this pr has a node expr that matches 'the_node).
* The observed behavior means that a Resource Override is evaluation order dependent (a similar evaluation order problem can be constructed without involving the ENC). 
* What is the expected behavior when using both ENC and node expressions to classify?

(Think this discussion is best had on the ticket).